contracts/contracts/token/OUSD.sol

```diff
@@ -445,159 +441,258 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
-
-                // Set fixed credits per token for this account
-                nonRebasingCreditsPerToken[_account] = _rebasingCreditsPerToken;
-                // Update non rebasing supply
-                nonRebasingSupply = nonRebasingSupply.add(balanceOf(_account));
-                // Update credit tallies
-                _rebasingCredits = _rebasingCredits.sub(
-                    _creditBalances[_account]
-                );
-            }
-        }
+    function _balanceToRebasingCredits(uint256 _balance)
+        internal
+        view
+        returns (uint256 rebasingCredits)
+    {
+        // Rounds up, because we need to ensure that accounts always have
+        // at least the balance that they should have.
+        // Note this should always be used on an absolute account value,
+        // not on a possibly negative diff, because then the rounding would be wrong.
+        return  ((_balance) * rebasingCreditsPerToken_ + 1e18 - 1) / 1e18;
     }
 
     /**
-     * @notice Enable rebasing for an account.
-     * @dev Add a contract address to the non-rebasing exception list. The
-     * address's balance will be part of rebases and the account will be exposed
-     * to upside and downside.
+     * @notice The calling account will start receiving yield after a successful call.
      * @param _account Address of the account.
      */
-    function governanceRebaseOptIn(address _account)
-        public
-        nonReentrant
-        onlyGovernor
-    {
+    function governanceRebaseOptIn(address _account) external onlyGovernor {
+        require(_account != address(0), "Zero address not allowed");
         _rebaseOptIn(_account);
     }
 
     /**
-     * @dev Add a contract address to the non-rebasing exception list. The
-     * address's balance will be part of rebases and the account will be exposed
-     * to upside and downside.
+     * @notice The calling account will start receiving yield after a successful call.
      */
-    function rebaseOptIn() public nonReentrant {
+    function rebaseOptIn() external {
         _rebaseOptIn(msg.sender);
     }
 
     function _rebaseOptIn(address _account) internal {
-        require(_isNonRebasingAccount(_account), "Account has not opted out");
-
-        // Convert balance into the same amount at the current exchange rate
-        uint256 newCreditBalance = _creditBalances[_account]
-            .mul(_rebasingCreditsPerToken)
-            .div(_creditsPerToken(_account));
+        uint256 balance = balanceOf(_account);
 
-        // Decreasing non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.sub(balanceOf(_account));
-
-        _creditBalances[_account] = newCreditBalance;
+        // prettier-ignore
+        require(
+            alternativeCreditsPerToken[_account] > 0 ||
+                // Accounts may explicitly `rebaseOptIn` regardless of
+                // accounting if they have a 0 balance.
+                balance == 0
```

Comment:
After decoding the certora ping on this method, I've tracked it down to bad accounting created when account credits are > 0 but < rebasingCreditsPerToken.  I'm not sure that this can actually happen in real life, however, if it does then the already rebasing credits on that are already account are not accounted for when updating the globals.

1. Do we need the balance==0 check at all?
2. If so, can we replace it with a `creditBalances[_account] == 0` check, which should be a stronger check?


> Me to certora.
>
> Looks like the exampe is a rebasing account that is opting into being a rebasing account. It's skipping the accounting check because the balance is zero. However that zero balance in the example is coming from having some credits on the account that are divided by a larger global rebasingCreditsPerToken, resulting in 0. 
>
> I'm not sure this is a situation that could actually happen in practice, since in order to have any credits at all, then the account would have had to been transfered or minted >= 1 raw balance. If they had 1 raw balance, then they would have had account credits = the global rebasingCreditsPerToken. Since global rebasingCreditsPerToken only goes down, it shouldn't be possible in real life to have a rebasing account with (account credits < rebasingCreditsPerToken AND account credits > 0) and thus should not be possible to have an stdRebasing account with credits but not a balance.
>
> Nevertheless, we'll probably the code so this rule would pass successfully as the rule is currently written. Given a choice, I'd rather have the code be correct in all circumstances, not just all possible circumstances.

Reply:
Test reproduction: 

```
        ousd.mint(user, 1);
        ousd.changeSupply(ousd.totalSupply()-1e18);
        vm.prank(user);
        ousd.rebaseOptIn();
```

Reply:
I just temporarily altered the check to use `creditBalances[_account] == 0`, so that we can see with certora if there are any other bad behaviors out of this function.

Reply:
Yes agreed, the `creditBalance` check should be a stronger one and still perform its old duty by allowing an empty rebasing account to explicitly optIn into rebasing.

I think your finding is correct that this failed state can only be reached if one rebases negatively consequently changing the supply downwards.

The storage data from the tests suggest the same with really large numbers for `rebasingCreditsPerToken`:
<img width="399" alt="Screenshot 2024-12-17 at 23 16 20" src="https://github.com/user-attachments/assets/31438437-4193-407c-acac-529086532a14" />

Reply:
@sparrowDom What was the reason for allowing zero balance rebaseOptIns?

Reply:
Yes that is the only reason I can think of. We do allow calling `rebaseOptIn` for EOA accounts with 0 balance as well, though I see no real utility there.